### PR TITLE
Use pp-sketchlib for lineage assignment code

### LINKED ISSUE #79: Make lineage assignment into a model object
Move the existing code so that it follows the inheritance structure of the other models.
Changing the command line options to give model type when using `--fit-model` would also make sense.

### PR DESCRIPTION
**Pull request by johnlees**
This updates the lineage assignment mode as follows:

- Makes the Lineage fit into a model object, the same as BGMM, DBSCAN and Refine
- Combine some of the main() control given that the code is now more similar between modes. The main difference is that lineage assign loops over ranks, and also outputs a combined lineage CSV (as before)
- Adds save, load, plot, fit and assign methods to this
- Fit gets nearest neighbours by passing the distance matrix to `pp_sketchlib.sparsifyDists()` with the appropriate rank, getting a sparse COO matrix in return
- Assign simply gets the i, j elements of the COO matrix, which can be used to directly construct the network
- Save and load store the rank in the pickle, and the sparse matrix in the npz
- Plot creates a histogram of the distances for the included links

Still to do is assign query in this mode, which will:

- [x] Load the sparse matrix
- [x] Calculate the q v q and q v r dists as usual
- [x] Add these into the sparse matrix, forming a dense section, but still as a sparse representation
- [x] Reapply the rank cutoff
- [x] Construct the network, but use the previous clusters to ensure names match correctly

Closes #79

### CODE REVIEW

**Comment by johnlees** [comment]
In theory this is all done, but needs testing (properly)

File: PopPUNK/__main__.py [hunk 1]
```diff
@@ -872,9 +863,34 @@ def assign_query(dbFuncs, ref_db, q_files, output, update_db, full_db, distances
...
         qcPass = qcDistMat(distMat, refList, queryList, max_a_dist)
 
-        # Assign to strains or lineages, as requested
+        # Assign to strains or lineages, as requested.
+        # Both need the previous model loaded
+        model_prefix = ref_db
+        if model_dir is not None:
+            model_prefix = model_dir
+        model_file = model_prefix + "/" + os.path.basename(model_prefix)
+
```
Comment (nickjcroucher) [inline PopPUNK/__main__.py:872]: Hit an error when querying - first problem looks to be that each model is saved with the rank number at the end, but model file loading does not append the rank number.

**Review by nickjcroucher — COMMENTED** [review]
Works very nicely in the database creation mode, suggested modifications are:

- Querying does not work at the moment - not finding the right database file - I think it is not appending the rank number to the model fit number?

- Would be nice to parallelise/iterate queries at different R

- Visualisation only outputs a single rank, not all clusterings and the combined cluster that is present in the full `lineages.csv`

File: PopPUNK/__main__.py [hunk 2]
```diff
@@ -604,37 +600,34 @@ def main():
...
+                           os.path.basename(args.output) + '_graph.gt',
+                           fmt = 'gt')
+
         #******************************#
         #*                            *#
         #* clique pruning             *#
         #*                            *#
         #******************************#
         # extract limited references from clique by default
+        # (this no longer loses information and should generally be kept on)
```
[the remark this answers is not in the document]
Comment (johnlees) [reply PopPUNK/__main__.py:613]: This will be dealt with in #112

File: PopPUNK/__main__.py [hunk 3]
```diff
@@ -315,10 +310,11 @@ def main():
...
             sys.stderr.write('Ranks should be small non-zero integers for sensible results\n')
-            exit(1)
-        if args.assign_lineages and args.existing_scheme is None:
-            sys.stderr.write('Must provide an existing scheme (--existing-scheme) if assigning to lineages\n')
-            exit(1)
+            sys.exit(1)
+        if args.assign_lineages:
+            if len(rank_list) > 1:
+                sys.stderr.write('Only assigning using lowest provided rank')
+            rank_list = min(rank_list)
```
[the remark this answers is not in the document]
Comment (johnlees) [reply PopPUNK/__main__.py:318]: Now moved iteration into the model object, which should also deal with the file name issues

File: PopPUNK/reference_pick.py [hunk 4]
```diff
@@ -86,12 +81,14 @@ def main():
     G_ref.save(args.output + "/" + os.path.basename(args.output) + '_graph.gt', fmt = 'gt')
 
     # Prune distances
+    nodes_to_remove = set(range(len(refList))).difference(reference_indices)
+    names_to_remove = [refList[n] for n in nodes_to_remove]
     prune_distance_matrix(refList, nodes_to_remove, distMat,
                           args.output + "/" + os.path.basename(args.output) + ".dists")
```
[the remark this answers is not in the document]
Comment (johnlees) [reply PopPUNK/reference_pick.py:88]: Deal with this in #110 or #112

**Pull request merged**